File: Queue/linked_queue.py
#!/usr/bin/env python

# An implementation of queue using linked list


class ListNode:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


class LinkedQueue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def get_size():
        return self.size

    def is_empty():
        return self.size == 0

    def enqueue(item):
        node = ListNode(item)
        if not self.head:
            self.head = node
        else:
            self.tail.next = node

        self.tail = node
        self.size += 1

    def dequeue(self):
        value = self.head.value
        if self.head == self.tail:
            self.tail = None
        self.head = self.head.next
        self.size -= 1
        return value

File: Queue/LinkedQueue.java
import java.util.NoSuchElementException;

/**
 * An implementation of queue using linked list.
 */

public class LinkedQueue<T> implements Queue<T> {
    private class ListNode {
        public T value;
        public ListNode next;

        public ListNode(T value, ListNode next) {
            this.value = value;
            this.next = next;
        }
    }

    private int size;
    private ListNode head;
    private ListNode tail;

    public int size() { return size; }

    public boolean isEmpty() { return size == 0; }

    public void enqueue(T item) {
        ListNode node = new ListNode(item, null);

        if (head == null) {
            head = node;
        } else {
            tail.next = node;
        }

        tail = node;
        size++;
    }

    public T dequeue() {
        T value = head.value;

        if (head == tail) {
            tail = null;
        }

        head = head.next;
        size--;

        return value;
    }

    public T peek() { return head.value; }

    public static void main(String[] args) {
        LinkedQueue<Integer> queue = new LinkedQueue<Integer>();
        System.out.println(queue.size());
        System.out.println(queue.isEmpty());
        for (int i = 0; i < 10; i++) {
            queue.enqueue(i);
        }
        for (int i = 0; i < 10; i++) {
            System.out.print(queue.dequeue() + " ");
        }
    }
}
